src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,295 @@
+#include "SpatialJoin.h"
+#include "util/AllocatorWithLimit.h"
+#include "util/MemorySize/MemorySize.h"
+#include "global/ValueId.h"
+#include "ValuesForTesting.h"
+#include "parser/ParsedQuery.h"
+#include "VariableToColumnMap.h"
+#include "engine/ExportQueryExecutionTrees.h"
+#include "util/GeoSparqlHelpers.h"
+
+SpatialJoin::SpatialJoin(QueryExecutionContext* qec, SparqlTriple triple,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childLeft,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childRight,
+  int maxDist)
+            : Operation(qec) {
+  this->triple = triple;
+  this->maxDist = maxDist;
+  if (childLeft) {
+    this->childLeft = childLeft.value();
+  }
+  if (childRight) {
+    this->childRight = childRight.value();
+  }
+  
+  if (triple._s.isVariable() && triple._o.isVariable()) {
+    leftChildVariable = triple._s.getVariable();
+    rightChildVariable = triple._o.getVariable();
+  } else {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+}
+
+std::shared_ptr<SpatialJoin> SpatialJoin::addChild(
+      std::shared_ptr<QueryExecutionTree> child, Variable varOfChild){
+  if (varOfChild == leftChildVariable) {
+    childLeft = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else if (varOfChild == rightChildVariable) {
+    childRight = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else {
+    AD_THROW("variable does not match");
```

Comment:
Currently this does not yet work as intended:

The query planner must not only check if the spatial join is fully constructed, but if the Variable that it wants to join on, is still available.
Edit: I'll have to check, this might work if you have your VariableToColumnMaps setup accordintly. 

In any case you should start writing unit tests.